mkosi/addon.py

```diff
@@ -0,0 +1,151 @@
+
+import mkosi.resources
+from mkosi.config import DocFormat, OutputFormat
+from mkosi.documentation import show_docs
+from mkosi.log import log_notice, log_setup
+from mkosi.run import run, uncaught_exception_handler
+from mkosi.sandbox import __version__, umask
+from mkosi.tree import copy_tree
+from mkosi.types import PathString
+from mkosi.util import initrd_copy_host_config, resource_path
+
+
+@uncaught_exception_handler()
+def main() -> None:
+    log_setup()
+
+    parser = argparse.ArgumentParser(
+        prog="mkosi-addon",
+        description="Build initrd/cmdline/ucode addon for the current system using mkosi",
+        allow_abbrev=False,
+        usage="mkosi-addon [options...]",
+    )
+
+    parser.add_argument(
+        "--kernel-version",
+        metavar="KERNEL_VERSION",
+        help="Kernel version string",
+        default=platform.uname().release,
+    )
+    parser.add_argument(
+        "-o",
+        "--output",
+        metavar="NAME",
+        help="Output name",
+        default="mkosi-local.addon.efi",
+    )
+    parser.add_argument(
+        "-O",
+        "--output-dir",
+        metavar="DIR",
+        help="Output directory",
+        default="",
+    )
+    parser.add_argument(
+        "--debug",
+        help="Turn on debugging output",
+        action="store_true",
+        default=False,
+    )
+    parser.add_argument(
+        "--debug-shell",
+        help="Spawn debug shell if a sandboxed command fails",
+        action="store_true",
+        default=False,
+    )
+    parser.add_argument(
+        "-D",
+        "--show-documentation",
+        help="Show the man page",
+        action="store_true",
+        default=False,
+    )
+    parser.add_argument(
+        "--version",
+        action="version",
+        version=f"mkosi {__version__}",
+    )
+
+    args = parser.parse_args()
+
+    if args.show_documentation:
+        with resource_path(mkosi.resources) as r:
+            show_docs("mkosi-addon", DocFormat.all(), resources=r)
+        return
+
+    # No local configuration? Then nothing to do
+    if not Path("/etc/mkosi-addon").exists() and not Path("/run/mkosi-addon").exists():
+        return
+
+    with tempfile.TemporaryDirectory() as staging_dir:
+        cmdline: list[PathString] = [
+            "mkosi",
+            "--force",
+            "--directory", "",
+            f"--format={str(OutputFormat.addon)}",
+            "--output", args.output,
+            "--output-directory", staging_dir,
+            "--build-sources", "",
+            "--include=mkosi-addon",
+        ]  # fmt: skip
+
+        cmdline += [
+            "--extra-tree=/usr/lib/systemd/boot/efi:/usr/lib/systemd/boot/efi",
+            "--remove-files=/usr/lib/systemd/boot/efi/",
+            "--sandbox-tree",
+            f"/usr/lib/modules/{args.kernel_version}:/usr/lib/modules/{args.kernel_version}",
+            "--sandbox-tree=/usr/lib/firmware:/usr/lib/firmware",
+        ]  # fmt: skip
```

Comment:
Just combine these, and move the systemd-stub stuff into the config file

Reply:
Also the sandbox trees for modules and firmware should be extra trees

Reply:
extra-tree means the entire thing gets included - all the firmware and all the kernel modules

Reply:
Not if you add `KernelModulesExclude=.*`

Reply:
And then other stuff can be selectively included

Reply:
`KernelModulesExclude=.*` does not work, everything is included anyway

Reply:
https://github.com/systemd/mkosi/pull/3381

Reply:
Still doesn't work with that either